README
README for Jabber Open Source Server (2.x)

Thanks for downloading jabberd2. Below are some basic instructions to
get you started. Complete documentation is available at http://jabberd2.org/

-- the jabberd team


Required packages:

 - expat - XML parsing libraries
     http://expat.sourceforge.net/
 - GnuSASL (1.1 or higher) - Simple Authentication and Security Layer library
     http://www.gnu.org/software/gsasl/
 - UDNS - asynchronous DNS resolver library
     http://www.corpit.ru/mjt/udns.html

Optional packages:

 - GNU Libidn (0.3.0 or higher) - needed for JID canonicalisation
     http://www.gnu.org/software/libidn/
 - OpenSSL (0.9.6b or higher) - needed for SSL/TLS support
     http://www.openssl.org/news/
 - zlib (1.2.3 or higher) - needed for stream compression
     http://www.zlib.net/
 - Berkeley DB (4.1.24 or higher)
     http://www.sleepycat.com/download/
 - OpenLDAP (2.1.0 or higher)
     http://www.openldap.org/software/download/
 - PostgresSQL (8.0 or higher; development libraries and headers)
     http://www.postgresql.org/
 - MySQL (5.0 or higher; development libraries and headers)
     http://www.mysql.com/
 - PAM
     http://www.linux-pam.org/  (for Linux)
 - SQLite (3.0 or higher)
     http://www.sqlite.org/


Build:

  % ./configure
  % make
  % make install

Options to ./configure:

  --prefix=/path/to/install
      Where to install everything

  --enable-idn (default: enabled)
      Compile with IDN (Stringprep) support

  --enable-ssl (default: enabled)
      Compile with SSL/TLS support

  --enable-mysql (default: enabled)
      Compile MySQL auth/reg/storage support

  --enable-pgsql (default: disabled)
      Compile PostgreSQL auth/reg/storage support

  --enable-db (default: disabled)
      Compile Berkeley DB auth/reg/storage support

  --enable-sqlite (default: disabled)
      Compile SQLite storage support

  --enable-ldap (default: disabled)
      Compile OpenLDAP auth/reg support

  --enable-pam (default: disabled)
      Compile PAM auth/reg support

  --enable-pipe (default: disabled)
      Compile pipe auth/reg support

  --enable-anon (default: disabled)
      Compile anonymous auth support

  --enable-fs (default: disabled)
      Compile filesystem storage support (NOT RECOMMENDED)

  --enable-mio=BACKEND
      Specify MIO backend to use (poll or select)

  --with-extra-include-path
  --with-extra-library-path
      Add extra header/library search paths
      
  --enable-debug (default: disabled)
      Output debugging info when run with -D

  --enable-nad-debug (default: disabled)
      Compile with NAD pointer tracking

  --enable-pool-debug (default: disabled)
      Compile with pool statistics

  --enable-developer (default: disabled)
      Compile with full warnings and debugging symbols


Configure:

  Edit $prefix/etc/(router|sm|c2s|resolver|s2s).xml to taste. In
  particular, make sure you setup for your choice of data storage
  correctly. If you're using the Berkeley DB backend, you'll need to
  create /var/run/jabberd and sets its permissions so that the server
  processes can find it.
  
  If you're using a SQL backend, you'll need to create an account for
  the server to use, and create the tables. Load db-setup.mysql or
  db-setup.pgsql from the tools/ directory into your database to do
  this.

  If you plan to use the jabberd wrapper script, make sure you look at
  the paths in the $prefix/etc/jabber/jabberd.cfg.


Run:

  You can either run all of the pieces separately:

    % $prefix/bin/router &
    % $prefix/bin/s2s &
    % $prefix/bin/sm &
    % $prefix/bin/c2s &

  Or you can run them all from the jabberd wrapper script:

    % $prefix/jabberd &

All the processes can take the following switches:

    -c use an alternate config file
    -D output lots of debugging info (if compiled with --enable-debug)


Upgrade:

  Please see NEWS file.


Support:

Webpage: http://jabberd2.org/
Mailinglist: jabberd2@lists.xiaoka.com
 (Subscribe by sending mail to jabberd2-subscribe@lists.xiaoka.com)

When requesting assistance, please note that the following things can
provide useful information which may assist with finding your problem:

 - debug logs (compile with --enable-debug and run with -D)
 - running components seperately (ie without the wrapper script)
 - config.log

Please try to provide as much relevant information as possible when
reporting problems - it will make helping you much easier.


Copyright & License:
  
 jabberd - Jabber Open Source Server
 Copyright (c) 2002-2012 Jeremie Miller, Thomas Muldowney,
                         Ryan Eatmon, Robert Norris, Tomasz Sterna.

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA02111-1307USA

 As a special exception, the authors give permission to link this
 program with the OpenSSL library and distribute the resulting binary.


 subst/snprintf.c and util/base64.c were originally taken from the
 Apache web server project.
   Originally copyright (c) 1995-2003 Apache Software Foundation.
 
 util/md5.c was taken from Ghostscript.
   Originally copyright (c) 1999-2002 Aladdin Enterprises.
 
 util/sha1.c was taken from Mozilla.
   Originally copyright (c) 1995-1999 Cryptography Research, Inc.
 
 subst/getopt.[ch] was taken from GNU Libc.
   Originally copyright (c) 1987-1993 Free Software Foundation, Inc.
 
 subst/gettimeofday.c was taken from PostgreSQL.
   Originally copyright (c) 2003 SRA, Inc. & SKC, Inc.
 
 subst/syslog.[ch] was taken from Bind.
   Originally copyright (c) 2001 Internet Software Consortium.
 
 subst/inet_aton.c
   Originally copyright (c) 1995-1997 Kungliga Teniska Hogskolan
 
 subst/ip6_misc.h
   Originally copyright (c) 1993,1994,1997 The Regents of the University of California.
 
 subst/dirent.[ch]
   Originally copyright (c) 1997,2003 Kevlin Henney.


NEWS
This file contains news, important changes
and upgrade instructions between different versions of jabberd2.


* 2.6.1 to 2.7.0 upgrade:

What changed:
- openssl-1.1 support
- Python script to export data in XEP-0227 format
- Fix build errors with MariaDB 10.2
- systemd units restart on failure
- Use systemd BindsTo= instead of BindTo.


* 2.6.0 to 2.6.1 upgrade:

What changed:
- Fixed offered SASL mechanism check

This is a security bugfix release.


* 2.5.0 to 2.6.0 upgrade:

What changed:
- Better SASL error messages

This is a bugfix release.


* 2.4.0 to 2.5.0 upgrade:

What changed:
- Do not attempt to reload SM modules on SIGHUP
- Cleanup config files example
- Fixed memory leak in pgsql storage driver
- Fixed two double-frees caused by dangling pointers
- Fixed c2s logger initialization point

This is a bugfix release.
The main visible changes that SIGHUP to SM will logrotate only.


* 2.3.6 to 2.4.0 upgrade:

What changed:
- Check for C99 support in compiler
- Count RIO bytes and check against max stanza size
- Gracefully drop unhandled HTTP connections
- wss:// (WebSocket over SSL) support in c2s
- Allow BareJID S10N packets
- Fallback to connecting S2S using local.ip when none of the origin.ip works
- Removed explicit SQLite transactions
- SQLite postconnect SQL support
- SQLite DB setup script improvements
- Many Coverity Scan and cppcheck detected issues fixed
- Properly lowercase SASL mechanisms in c2s
- Support out-of-source build

This is a bugfix release.


* 2.3.5 to 2.3.6 upgrade:

What changed:
- Support WebSocket fragmented packets
- Fixed delivering directed presence (to self)
- Reset in-sess 'from' to FullJID on non-Presence packets

This is mainly a bugfix release.
The main change is that WebSocket connections work stable now.


* 2.3.4 to 2.3.5 upgrade:

What changed:
- Module to verify users using e-mail
- Reordered MIO backends priority
- Skip non-existing blowfish i386 assembler code
- Use CSPRNG for dialback keys
- Allow presence probing own connections
- Use OpenSSL functions for base64 en/decoding when available
- Option to dump packet-filter matched packets to file

mod_verify requires CREATE TABLE "verify" in DB. Make sure you
created it before enabling the module in sm.xml.

MIO backends are prioritized from best to worst now, so if you
do not enforce a backend with ./configure it may change
in new build.

jabberd2 is now leaning strongly against OpenSSL. It is still
possible to use without, but not advisable.


* 2.3.3 to 2.3.4 upgrade:

What changed:
- Rewrite TLS ephemeral key + cipher handling
- Recover Berkeley DB before opening it
- bcrypt support for PostgreSQL
- Option to set authreg module per realm
- AuthReg ANONYMOUS does not offer password check
- Answer to disco#info queries to user JID
- WebSocket C2S SX plugin

ANONYMOUS login was extended and fixed, so make sure it still
works for you if you were using it in its semi-broken state before.
See sm.xml for notes on configuring ANONYMOUS login and apply
accordingly.

Notice additional module in pkt-user chain in sm.xml.

C2S support for WebSocket connections is documented in c2s.xml.
Notice changes to HTTP-Forward option - it requires enabled
WebSocket support.


* 2.3.2 to 2.3.3 upgrade:

What changed:
- Support for RSA/DH/ECDH key agreement
- bcrypt support for MySQL storage
- C2S per session user data & authreg auth API extensions
  for custom authreg backends
- Option to provide a custom the openssl library path


* 2.3.1 to 2.3.2 upgrade:

What changed:
- Removed unmaintained CyrusSASL backend
- Option to add realm to username in ldapvcard module
- systemd unit files


* 2.3.0 to 2.3.1 upgrade:

What changed:
- Marked "TLS-Everywhere" as EXPERIMENTAL feature
- default EXPERIMENTAL to 'no'
- default SUPERSEDED to 'no'
- moved STANZA-ACK and MY-IP-ADDRESS XEPs and IQ-PRIVATE push
  out of experimental status


* 2.2.17 to 2.3.0 upgrade:

What changed:
- Renamed non-standard UPGRADE file overwriting outdated NEWS file
- Semantic Versioning: http://semver.org/
- TLS Everywhere: https://github.com/stpeter/manifesto
- Required GSASL >=1.1
- jabberd should compile without warnings
- out-of-source builds should work
- pgsql: authreg password_type support
- pgsql: schema support
- ldapvcard: groupattr works even if no groupattr_regex defined
- ldapfull: checks for ldap group membership on login
- vCard: Assume tel phone is voice phone
- MySQL: default password hashing algorithm changed to SHA512
- out-conn-reuse s2s.xml option naming unified
- XML parse error will log buffer details
- CRAM-MD5 auth support
- router private key cachain and password support
- hashed passwords support in SQLite3 storage


* 2.2.16 to 2.2.17 upgrade:

What changed:
- cppunit was replaced by check framework
- Fixed build with debug disabled


* 2.2.15 to 2.2.16 upgrade:

What changed:
- Debug logging into file implemented
  See etc/*.xml.dist for instructions
- Unit tests are not mandatory
  See --enable-tests/--disable-tests option for ./configure
- Fix for invalid default router.xml (Unterminated XML comment tag)


* 2.2.14 to 2.2.15 upgrade:

What changed:
- Merged config variables expansion
- Group extraction by RegExp from LDAP
- A1HASH passwords support for MySQL
- PBX integration pipe implementation
- Enable/disable debug output with signals USR1/USR2
- '-i ID' command line option
- DB query rate limitting
- Domain whitelisting support
- Support enforcing TLS secured S2S connections only
- Simple message logging to a directory
- MIO kqueue implementation
- Support for see-other-host stream redirects
- Unit tests

You will need cppunit(-devel) installed to get the tests built.
See example configuration files on how to enable new features.


* 2.2.13 to 2.2.14 upgrade:

What changed:
- There are example upstart configuration files in etc/ you may use.

* 2.2.12 to 2.2.13 upgrade:

sm/pkt.c
/*
 * jabberd - Jabber Open Source Server
 * Copyright (c) 2002 Jeremie Miller, Thomas Muldowney,
 *                    Ryan Eatmon, Robert Norris
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA02111-1307USA
 */

#include "sm.h"
#include "lib/stanza.h"
#include "lib/datetime.h"

/** @file sm/pkt.c
  * @brief packet abstraction
  * @author Robert Norris
  * $Date: 2005/09/09 05:34:13 $
  * $Revision: 1.35 $
  */

pkt_t *pkt_error(pkt_t *pkt, int err) {
    if(pkt == NULL) return NULL;

    /* if it's an error already, log, free, return */
    if(pkt->type & pkt_ERROR) {
//        LOG_DEBUG(sm->log, "dropping error pkt");
        pkt_free(pkt);
        return NULL;
    }

    stanza_error(pkt->nad, 1, err);

    /* update vars and attrs */
    pkt_tofrom(pkt);
    pkt->type |= pkt_ERROR;

    /* supplant route destination in case there was none in original packet */
    if(pkt->to == NULL && pkt->rto != NULL)
        pkt->to = jid_dup(pkt->rto);

    /* all done, error'd and addressed */
//    LOG_DEBUG(sm->log, "processed %d error pkt", err);

    return pkt;
}

/** swap a packet's to and from attributes */
pkt_t *pkt_tofrom(pkt_t *pkt) {
    jid_t *tmp;

    if(pkt == NULL) return NULL;

    /* swap vars */
    tmp = pkt->from;
    pkt->from = pkt->to;
    pkt->to = tmp;
    tmp = pkt->rfrom;
    pkt->rfrom = pkt->rto;
    pkt->rto = tmp;

    /* update attrs */
    if(pkt->to != NULL)
        nad_set_attr(pkt->nad, 1, -1, "to", jid_full(pkt->to), 0);
    if(pkt->from != NULL)
        nad_set_attr(pkt->nad, 1, -1, "from", jid_full(pkt->from), 0);
    if(pkt->rto != NULL)
        nad_set_attr(pkt->nad, 0, -1, "to", jid_full(pkt->rto), 0);
    if(pkt->rfrom != NULL)
        nad_set_attr(pkt->nad, 0, -1, "from", jid_full(pkt->rfrom), 0);

    return pkt;
}

/** duplicate pkt, replacing addresses */
pkt_t *pkt_dup(pkt_t *pkt, const char *to, const char *from) {
    pkt_t *pnew;

    if(pkt == NULL) return NULL;

    pnew = new(pkt_t);

    pnew->sm = pkt->sm;
    pnew->type = pkt->type;
    pnew->nad = nad_copy(pkt->nad);

    /* set replacement attrs */
    if(to != NULL) {
        pnew->to = jid_new(to, -1);
        nad_set_attr(pnew->nad, 1, -1, "to", jid_full(pnew->to), 0);
    } else if(pkt->to != NULL)
        pnew->to = jid_dup(pkt->to);

    if(from != NULL) {
        pnew->from = jid_new(from, -1);
        nad_set_attr(pnew->nad, 1, -1, "from", jid_full(pnew->from), 0);
    } else if(pkt->from != NULL)
        pnew->from = jid_dup(pkt->from);

//    LOG_DEBUG(sm->log, "duplicated packet");

    return pnew;
}

pkt_t *pkt_new(sm_t *sm, nad_t *nad) {
    pkt_t *pkt;
    int ns, attr, elem;
    char pri[20];

    LOG_DEBUG(sm->log, "creating new packet");

    /* find the route */
    ns = nad_find_namespace(nad, 0, uri_COMPONENT, NULL);
    if(ns < 0) {
        LOG_DEBUG(sm->log, "packet not in component namespace");
        nad_free(nad);
        return NULL;
    }

    /* create the pkt holder */
    pkt = new(pkt_t);

    pkt->sm = sm;
    pkt->nad = nad;

    /* routes */
    if(NAD_ENAME_L(nad, 0) == 5 && strncmp("route", NAD_ENAME(nad, 0), 5) == 0) {
        /* route element */
        if((attr = nad_find_attr(nad, 0, -1, "to", NULL)) >= 0)
            pkt->rto = jid_new(NAD_AVAL(nad, attr), NAD_AVAL_L(nad, attr));
        if((attr = nad_find_attr(nad, 0, -1, "from", NULL)) >= 0)
            pkt->rfrom = jid_new(NAD_AVAL(nad, attr), NAD_AVAL_L(nad, attr));

        /* route type */
        attr = nad_find_attr(nad, 0, -1, "type", NULL);
        if(attr < 0)
            pkt->rtype = route_UNICAST;
        else if(NAD_AVAL_L(nad, attr) == 9 && strncmp("broadcast", NAD_AVAL(nad, attr), 9) == 0)
            pkt->rtype = route_BROADCAST;

        /* route errors */
        if(nad_find_attr(nad, 0, -1, "error", NULL) >= 0)
            pkt->rtype |= route_ERROR;

        /* client packets */
        ns = nad_find_namespace(nad, 1, uri_CLIENT, NULL);
        if(ns >= 0) {

            /* get initial addresses */
            if((attr = nad_find_attr(pkt->nad, 1, -1, "to", NULL)) >= 0 && NAD_AVAL_L(pkt->nad, attr) > 0)
                pkt->to = jid_new(NAD_AVAL(pkt->nad, attr), NAD_AVAL_L(pkt->nad, attr));
            if((attr = nad_find_attr(pkt->nad, 1, -1, "from", NULL)) >= 0 && NAD_AVAL_L(pkt->nad, attr) > 0)
                pkt->from = jid_new(NAD_AVAL(pkt->nad, attr), NAD_AVAL_L(pkt->nad, attr));

            /* find type, if any */
            attr = nad_find_attr(pkt->nad, 1, -1, "type", NULL);

            /* messages are simple, only subtypes */
            if(NAD_ENAME_L(pkt->nad, 1) == 7 && strncmp("message", NAD_ENAME(pkt->nad, 1), 7) == 0) {
                pkt->type = pkt_MESSAGE;
                if(attr >= 0) {
                    if(NAD_AVAL_L(pkt->nad, attr) == 4 && strncmp("chat", NAD_AVAL(pkt->nad, attr), 4) == 0)
                        pkt->type = pkt_MESSAGE_CHAT;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 8 && strncmp("headline", NAD_AVAL(pkt->nad, attr), 8) == 0)
                        pkt->type = pkt_MESSAGE_HEADLINE;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 9 && strncmp("groupchat", NAD_AVAL(pkt->nad, attr), 9) == 0)
                        pkt->type = pkt_MESSAGE_GROUPCHAT;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 5 && strncmp("error", NAD_AVAL(pkt->nad, attr), 5) == 0)
                        pkt->type = pkt_MESSAGE | pkt_ERROR;
                }

                return pkt;
            }

            /* presence is a mixed bag, s10ns in here too */
            if(NAD_ENAME_L(pkt->nad, 1) == 8 && strncmp("presence", NAD_ENAME(pkt->nad, 1), 8) == 0) {
                pkt->type = pkt_PRESENCE;
                if(attr >= 0) {
                    if(NAD_AVAL_L(pkt->nad, attr) == 11 && strncmp("unavailable", NAD_AVAL(pkt->nad, attr), 11) == 0)
                        pkt->type = pkt_PRESENCE_UN;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 5 && strncmp("probe", NAD_AVAL(pkt->nad, attr), 5) == 0)
                        pkt->type = pkt_PRESENCE_PROBE;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 9 && strncmp("subscribe", NAD_AVAL(pkt->nad, attr), 9) == 0)
                        pkt->type = pkt_S10N;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 10 && strncmp("subscribed", NAD_AVAL(pkt->nad, attr), 10) == 0)
                        pkt->type = pkt_S10N_ED;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 11 && strncmp("unsubscribe", NAD_AVAL(pkt->nad, attr), 11) == 0)
                        pkt->type = pkt_S10N_UN;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 12 && strncmp("unsubscribed", NAD_AVAL(pkt->nad, attr), 12) == 0)
                        pkt->type = pkt_S10N_UNED;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 5 && strncmp("error", NAD_AVAL(pkt->nad, attr), 5) == 0)
                        pkt->type = pkt_PRESENCE | pkt_ERROR;
                }

                /* priority */
                if((elem = nad_find_elem(pkt->nad, 1, NAD_ENS(pkt->nad, 1), "priority", 1)) < 0)
                    return pkt;

                if(NAD_CDATA_L(pkt->nad, elem) <= 0 || NAD_CDATA_L(pkt->nad, elem) > 19)
                    return pkt;

                memcpy(pri, NAD_CDATA(pkt->nad, elem), NAD_CDATA_L(pkt->nad, elem));
                pri[NAD_CDATA_L(pkt->nad, elem)] = '\0';
                pkt->pri = atoi(pri);

                if(pkt->pri > 127) pkt->pri = 127;
                if(pkt->pri < -128) pkt->pri = -128;

                return pkt;
            }

            /* iq's are pretty easy, but also set xmlns */
            if(NAD_ENAME_L(pkt->nad, 1) == 2 && strncmp("iq", NAD_ENAME(pkt->nad, 1), 2) == 0) {
                pkt->type = pkt_IQ;
                if (attr < 0) {
                    LOG_ERROR(sm->log, "dropping iq without type");
                    pkt_free(pkt);
                    return NULL;
                }
                if (NAD_AVAL_L(pkt->nad, attr) == 6 && strncmp("result", NAD_AVAL(pkt->nad, attr), 6) == 0) pkt->type = pkt_IQ_RESULT;
                else if (NAD_AVAL_L(pkt->nad, attr) == 5 && strncmp("error", NAD_AVAL(pkt->nad, attr), 5) == 0) pkt->type = pkt_IQ | pkt_ERROR;
                else if (NAD_AVAL_L(pkt->nad, attr) == 3 && strncmp("set", NAD_AVAL(pkt->nad, attr), 3) == 0) pkt->type = pkt_IQ_SET;
                else if (NAD_AVAL_L(pkt->nad, attr) != 3 || strncmp("get", NAD_AVAL(pkt->nad, attr), 3)) {
                    LOG_ERROR(sm->log, "dropping iq with bad type \"%.*s\"", NAD_AVAL_L(pkt->nad, attr), NAD_AVAL(pkt->nad, attr));
                    pkt_free(pkt);
                    return NULL;
                }

                if(pkt->nad->ecur > 2 && (ns = NAD_ENS(pkt->nad, 2)) >= 0)
                    pkt->ns = (int) (long) xhash_getx(pkt->sm->xmlns, NAD_NURI(pkt->nad, ns), NAD_NURI_L(pkt->nad, ns));

                return pkt;
            }

            LOG_DEBUG(sm->log, "unknown client namespace packet");

            return pkt;
        }

        /* sessions packets */
        ns = nad_find_namespace(nad, 1, uri_SESSION, NULL);
        if(ns >= 0) {

            /* sessions */
            if(NAD_ENAME_L(pkt->nad, 1) == 7 && strncmp("session", NAD_ENAME(pkt->nad, 1), 7) == 0) {

                /* find action */
                attr = nad_find_attr(pkt->nad, 1, -1, "action", NULL);

                if(attr >= 0) {
                    if(NAD_AVAL_L(pkt->nad, attr) == 5 && strncmp("start", NAD_AVAL(pkt->nad, attr), 5) >= 0)
                        pkt->type = pkt_SESS;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 3 && strncmp("end", NAD_AVAL(pkt->nad, attr), 3) >= 0)
                        pkt->type = pkt_SESS_END;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 6 && strncmp("create", NAD_AVAL(pkt->nad, attr), 6) >= 0)
                        pkt->type = pkt_SESS_CREATE;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 6 && strncmp("delete", NAD_AVAL(pkt->nad, attr), 6) >= 0)
                        pkt->type = pkt_SESS_DELETE;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 7 && strncmp("started", NAD_AVAL(pkt->nad, attr), 7) >= 0)
                        pkt->type = pkt_SESS | pkt_SESS_FAILED;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 5 && strncmp("ended", NAD_AVAL(pkt->nad, attr), 5) >= 0)
                        pkt->type = pkt_SESS_END | pkt_SESS_FAILED;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 7 && strncmp("created", NAD_AVAL(pkt->nad, attr), 7) >= 0)
                        pkt->type = pkt_SESS_CREATE | pkt_SESS_FAILED;
                    else if(NAD_AVAL_L(pkt->nad, attr) == 7 && strncmp("deleted", NAD_AVAL(pkt->nad, attr), 7) >= 0)
                        pkt->type = pkt_SESS_DELETE | pkt_SESS_FAILED;

                    return pkt;
                } else {
                    LOG_DEBUG(sm->log, "missing action on session packet");
                    return pkt;
                }
            }

            LOG_DEBUG(sm->log, "unknown session namespace packet");

            return pkt;
        }

        LOG_DEBUG(sm->log, "unknown packet");

        return pkt;
    }

    /* advertisements */
    if(NAD_ENAME_L(nad, 0) == 8 && strncmp("presence", NAD_ENAME(nad, 0), 8) == 0) {
        if(nad_find_attr(nad, 0, -1, "type", "unavailable") >= 0)
            pkt->rtype = route_ADV_UN;
        else
            pkt->rtype = route_ADV;

        attr = nad_find_attr(nad, 0, -1, "from", NULL);
        if(attr >= 0)
            pkt->from = jid_new(NAD_AVAL(nad, attr), NAD_AVAL_L(nad, attr));

        return pkt;
    }

    LOG_DEBUG(sm->log, "invalid component packet");

    pkt_free(pkt);
    return NULL;
}

void pkt_free(pkt_t *pkt) {
//    LOG_DEBUG(sm->log, "freeing pkt");

    if (pkt != NULL) {
        if(pkt->rto != NULL) jid_free(pkt->rto);
        if(pkt->rfrom != NULL) jid_free(pkt->rfrom);
        if(pkt->to != NULL) jid_free(pkt->to);
        if(pkt->from != NULL) jid_free(pkt->from);
        if(pkt->nad != NULL) nad_free(pkt->nad);
        free(pkt);
    }
}

pkt_t *pkt_create(sm_t *sm, const char *elem, const char *type, const char *to, const char *from) {
    nad_t *nad;
    int ns;

    nad = nad_new();

    ns = nad_add_namespace(nad, uri_COMPONENT, NULL);
    nad_append_elem(nad, ns, "route", 0);

    nad_add_namespace(nad, uri_SESSION, "sm");

    ns = nad_add_namespace(nad, uri_CLIENT, NULL);
    nad_append_elem(nad, ns, elem, 1);

    if(type != NULL)
        nad_append_attr(nad, -1, "type", type);
    if(to != NULL)
        nad_append_attr(nad, -1, "to", to);
    if(from != NULL)
        nad_append_attr(nad, -1, "from", from);

    return pkt_new(sm, nad);
}

/** convenience - copy the packet id from src to dest */
void pkt_id(pkt_t *src, pkt_t *dest) {
    int attr;

    attr = nad_find_attr(src->nad, 1, -1, "id", NULL);
    if(attr >= 0)
        nad_set_attr(dest->nad, 1, -1, "id", NAD_AVAL(src->nad, attr), NAD_AVAL_L(src->nad, attr));
    else
        nad_set_attr(dest->nad, 1, -1, "id", NULL, 0);
}

/** create an id value for new iq packets */
void pkt_id_new(pkt_t *pkt) {
    char id[40];
    int i, r;

    /* as we are not using ids for tracking purposes, these can be generated randomly */
    for(i = 0; i < 40; i++) {
        r = (int) (36.0 * rand() / RAND_MAX);
        id[i] = (r >= 0 && r <= 9) ? (r + 48) : (r + 87);
    }

    nad_set_attr(pkt->nad, 1, -1, "id", id, 40);

    return;
}

void pkt_router(pkt_t *pkt) {
    mod_ret_t ret;
    int ns, scan;

    if(pkt == NULL) return;

//    LOG_DEBUG(sm->log, "delivering pkt to router");

    if(pkt->to == NULL) {
//        LOG_DEBUG(sm->log, "no to address on packet, unable to route");
        pkt_free(pkt);
        return;
    }

    if(pkt->rto != NULL)
        jid_free(pkt->rto);
    pkt->rto = jid_new(pkt->to->domain, -1);

    if(pkt->rto == NULL) {
//        LOG_DEBUG(sm->log, "invalid to address on packet, unable to route");
        pkt_free(pkt);
        return;
    }

    nad_set_attr(pkt->nad, 0, -1, "to", pkt->rto->domain, 0);

    if(pkt->rfrom != NULL)
        jid_free(pkt->rfrom);
    pkt->rfrom = jid_new(pkt->sm->id, -1);

    if(pkt->rfrom == NULL) {
//        LOG_DEBUG(sm->log, "invalid from address on packet, unable to route");
        pkt_free(pkt);
        return;
    }

    nad_set_attr(pkt->nad, 0, -1, "from", pkt->rfrom->domain, 0);

    ret = mm_out_router(pkt->sm->mm, pkt);
    switch(ret) {
        case mod_HANDLED:
            return;

        case mod_PASS:
            
            /* remove sm specifics */
            ns = nad_find_namespace(pkt->nad, 1, uri_SESSION, NULL);
            /* remove them if there is no session elements in packet */
            if(ns >= 0 && nad_find_elem(pkt->nad, 0, ns, NULL, 1) < 0) {
                nad_set_attr(pkt->nad, 1, ns, "c2s", NULL, 0);
                nad_set_attr(pkt->nad, 1, ns, "sm", NULL, 0);

                /* forget about the internal namespace too */
                if(pkt->nad->elems[1].ns == ns)
                    pkt->nad->elems[1].ns = pkt->nad->nss[ns].next;

                else {
                    for(scan = pkt->nad->elems[1].ns; pkt->nad->nss[scan].next != -1 && pkt->nad->nss[scan].next != ns; scan = pkt->nad->nss[scan].next);

                    /* got it */
                    if(pkt->nad->nss[scan].next != -1)
                        pkt->nad->nss[scan].next = pkt->nad->nss[ns].next;
                }
            }

            sx_nad_write(pkt->sm->router, pkt->nad);

            /* nad already free'd, free the rest */
            pkt->nad = NULL;
            pkt_free(pkt);

            break;

        default:
            pkt_router(pkt_error(pkt, -ret));

            break;
    }
}

void pkt_sess(pkt_t *pkt, sess_t *sess) {
    mod_ret_t ret;

    if(pkt == NULL) return;

    LOG_DEBUG(sess->user->sm->log, "delivering pkt to session %s", jid_full(sess->jid));

    if(pkt->rto != NULL)
        jid_free(pkt->rto);
    pkt->rto = jid_new(sess->c2s, -1);

    if(pkt->rto == NULL) {
        LOG_DEBUG(sess->user->sm->log, "invalid to address on packet, unable to route");
        pkt_free(pkt);
        return;
    }

    nad_set_attr(pkt->nad, 0, -1, "to", pkt->rto->domain, 0);

    if(pkt->rfrom != NULL)
        jid_free(pkt->rfrom);
    pkt->rfrom = jid_new(pkt->sm->id, -1);

    if(pkt->rfrom == NULL) {
        LOG_DEBUG(sess->user->sm->log, "invalid from address on packet, unable to route");
        pkt_free(pkt);
        return;
    }

    nad_set_attr(pkt->nad, 0, -1, "from", pkt->rfrom->domain, 0);

    ret = mm_out_sess(pkt->sm->mm, sess, pkt);
    switch(ret) {
        case mod_HANDLED:
            return;

        case mod_PASS:
            sess_route(sess, pkt);

            break;

        default:
            pkt_router(pkt_error(pkt, -ret));

            break;
    }
}

/** add an x:delay stamp */
void pkt_delay(pkt_t *pkt, time_t t, const char *from) {
    char timestamp[21];
    int ns, elem;

#ifdef ENABLE_SUPERSEDED
    datetime_out(t, dt_LEGACY, timestamp, 18);
    ns = nad_add_namespace(pkt->nad, uri_DELAY, NULL);
    elem = nad_insert_elem(pkt->nad, 1, ns, "x", NULL);
    nad_set_attr(pkt->nad, elem, -1, "stamp", timestamp, 0);
    if(from != NULL)
        nad_set_attr(pkt->nad, elem, -1, "from", from, 0);
//    LOG_DEBUG(sm->log, "added pkt XEP-0091 delay stamp %s", timestamp);
#endif
    datetime_out(t, dt_DATETIME, timestamp, 21);
    ns = nad_add_namespace(pkt->nad, uri_URN_DELAY, NULL);
    elem = nad_insert_elem(pkt->nad, 1, ns, "delay", NULL);
    nad_set_attr(pkt->nad, elem, -1, "stamp", timestamp, 0);
    if(from != NULL)
        nad_set_attr(pkt->nad, elem, -1, "from", from, 0);
//    LOG_DEBUG(sm->log, "added pkt XEP-0203 delay stamp %s", timestamp);
}
